src/libraries/Common/src/Interop/Windows/BCrypt/BCryptAlgorithmCache.cs

```diff
@@ -43,6 +44,44 @@ public static unsafe SafeBCryptAlgorithmHandle GetCachedBCryptAlgorithmHandle(st
                     }
                 }
             }
+
+            public static unsafe bool IsBCryptAlgorithmSupported(string hashAlgorithmId, BCryptOpenAlgorithmProviderFlags flags)
+            {
+                var key = (hashAlgorithmId, flags);
+
+                while (true)
```

Comment:
Why does this need looping?  Anyone who makes it past TryGetValue is going to call into OpenProv, and one of them (randomly) will write a success value.

If we're afraid that there will be too many calls to OpenProv, resulting in spurious failures, then the whole thing needs to be locked (since it could randomly result in the process agreeing on failure when it equivalently could agree on success).  Assuming we aren't afraid of that, all parallel callers should reach the same conclusion, and then every caller exits in either 1 (already cached) or 2 (not already cached: TryGetValue + TryAdd) bucket searches without the `while`, but 1 or 3 with it.

Reply:
It's modeled off of the get cached handle implementation in the same file, and it seemed reasonable to me. I can refactor it though to get rid of the loop.